### PR TITLE
ModalRoute resumes previous focus on didPopNext

### PR DESCRIPTION
**Pull request by dkwingsmt**
## Description

This PR changes `ModalRoute` so that, when the route on top of it is popped, it requests to focus on the last focus.

Currently when the user is editing something, the edit will be interrupted by popups or dialogs. After this PR, the edit will be resumed after the popup or dialog is closed. See the following example:

Before:
before

After:
after

## Related Issues

None

## Tests

I added the following tests:
- When a ModalRoute finds the route on top of it is popped, it resumes its last focus.
  - In the test, instead of ModalRoute and Route, we used PageRoute and DialogRoute, which might leave a risk, but otherwise we would have to build a ton of things from scratch.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Review by muttalipkucuk — APPROVED** [review]

File: packages/flutter/test/widgets/routes_test.dart
Comment (Piinks) [inline packages/flutter/test/widgets/routes_test.dart]: Can this be considered a `ModalRoute` test if it is using a `PageRoute` workaround?

**Comment by goderbauer** [comment]
@gspencergoog Can you take a look at this since you recently worked with the focus sub-system?

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

File: packages/flutter/test/widgets/routes_test.dart
Comment (dkwingsmt) [reply packages/flutter/test/widgets/routes_test.dart]: Yeah good point. I think the idea is that the code that it's testing is in ModalRoute. It's still a "ModalRoute test", it's just not good enough.

**Comment by tvolkert** [comment]
This change was identified as causing tests to fail inside Google.  Since other teams are blocked on the roll, we'll have to investigate tomorrow what the underlying cause is (it's possible that the test is doing something wrong, or it's possible there's a bug in this PR).  For now, I'm going to revert to unblock the roll.

**Pull request merged**
